source/agora/consensus/Reward.d

```diff
@@ -0,0 +1,410 @@
+/*******************************************************************************
+
+    Contains functions to calculate the block reward.
+
+    Block reward are given to validators for verifying transactions and signing
+    blocks. This serves as an incentive for coin holders to participate in the
+    network as validators. The more active validators the network has, the more
+    signatures are on the blocks, and ultimately the more secure the overall
+    network is.
+
+    Block reward is also sent to the Commons Budget address for the first
+    6.7 years. After that period, the Commons Budget projects will be funded
+    from the transaction fees.
+
+    Here are some finer details of the block reward calculation
+
+    $(UL
+
+    $(LI Block rewards will be distributed in every Coinbase payout block which
+    is also used to payout the transaction fees and data fees.)
+
+    $(LI The distribution of the block rewards will be delayed by one payout
+    period as is the case for fees. The rational behind this is that enough
+    time must be given to validators to share block signatures.)
+
+    $(LI Each validator will be rewarded based on the percentage of the
+    actual / maximum possible block signatures for the payout period.)
+
+    $(LI The total block reward for a particular height is fixed, however this
+    doesn't mean that the total validator reward is fixed. The total validator
+    reward is based on the total number of signatures on the blocks in question.
+    The total validator reward is reduced using a multiplier function of the
+    missing signatures, meaning that the more that are missing the higher the
+    penalty per missing signature. This is to encourage validators to share
+    signatures.
+    The coins remaining will go to the Commons Budget.
+    i.e. `fixed block reward at height X - total validator reward`)
+
+    $(LI The part of the total validator reward that cannot be divived equally
+    between the validators will also go to the Commons Budget.)
+
+    )
+
+    Copyright:
+        Copyright (c) 2019-2021 BOSAGORA Foundation
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.consensus.Reward;
+
+import agora.common.Amount;
+import agora.common.Ensure;
+import agora.common.Types;
+import agora.utils.Log;
+
+import std.algorithm;
+import std.array;
+import std.math;
+import std.range;
+
+mixin AddLogger!();
```

Comment:
Move it inside of the class

Reply:
Updated, although had to remove quite a few attributes in the class functions.